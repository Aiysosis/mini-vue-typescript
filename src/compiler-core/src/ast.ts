import {
	ASTNode,
	ElementNode,
	ExpressionNode,
	InterPolationNode,
	TextNode,
} from "./parse";

export enum NodeTypes {
	INTERPOLATION,
	SIMPLE_EXPRESSION,
	ELEMENT,
	TEXT,
	ROOT,
}

export const isElementNode = (node: ASTNode): node is ElementNode =>
	node.type === NodeTypes.ELEMENT;

export const isTextNode = (node: ASTNode): node is TextNode =>
	node.type === NodeTypes.TEXT;

export const isInterpolationNode = (node: ASTNode): node is InterPolationNode =>
	node.type === NodeTypes.INTERPOLATION;

export const isExpressionNode = (node: ASTNode): node is ExpressionNode =>
	node.type === NodeTypes.SIMPLE_EXPRESSION;
