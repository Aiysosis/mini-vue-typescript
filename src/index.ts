// export { computed } from "./reactivity/computed";

// export { ref, isRef, unRef } from "./reactivity/ref";

export { createRenderer } from "./runtime-core/index";
