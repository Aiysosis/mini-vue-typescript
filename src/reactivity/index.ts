export { computed } from "./computed";
export { effect } from "./effect";
export { isReactive, reactive } from "./reactive";
export { isRef, ref } from "./ref";
