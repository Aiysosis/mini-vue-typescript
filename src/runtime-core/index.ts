export { createRenderer } from "./renderer";
export { getCurrentInstance } from "./component";
export { createTextVNode } from "./vnode";
export { renderSlots } from "./helpers/renderSlots";
export { h } from "./vnode";
