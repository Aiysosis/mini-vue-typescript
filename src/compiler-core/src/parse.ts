import { NodeTypes } from "./ast";

const enum TagType {
	START,
	END,
}

export interface ASTNode {
	type: NodeTypes;
}

export interface ASTRoot extends ASTNode {
	codegenNode: ASTNode;
	helpers: symbol[];
	children: ASTNode[];
}

export interface ElementNode extends ASTNode {
	tag: string;
	children: ASTNode[];
}

export interface TextNode extends ASTNode {
	content: string;
}

export interface ExpressionNode extends ASTNode {
	content: string;
}

export interface InterPolationNode extends ASTNode {
	content: ExpressionNode;
}

export type CompoundExpressionChild = string | TextNode | InterPolationNode;

export class CompoundExpressionNode implements ASTNode {
	children: CompoundExpressionChild[];
	type = NodeTypes.COMPOUND_EXPRESSION;
	constructor(val?: CompoundExpressionChild) {
		this.children = [val];
	}
	push(this: CompoundExpressionNode, val: CompoundExpressionChild) {
		this.children.push(val);
	}
}

type ParseContext = {
	source: string;
};

export function baseParse(content: string) {
	const context = createParserContext(content);
	//* 解析的过程是从头到尾逐步推进的，所以需要解析一部分然后就剪掉，这样解析的过程才是可持续的。
	//* 而判断用什么方法进行解析，则是根据当前待解析字符串的头部元素进行判断，如如果发现字符串头部为 {{，
	//* 说明接下来是插值，那么就调用相应的函数进行解析。同时，用于步进剪切的函数可以抽离出来，这样各种类型的解析都可以复用

	return createRoot(parseChildren(context, []));
}

//@fn parseChildren
function parseChildren(context: ParseContext, ancesters: ElementNode[]) {
	const nodes: ASTNode[] = [];

	while (!isEnd(context, ancesters)) {
		let node = null;
		const s = context.source;

		if (s.startsWith("{{")) {
			node = parseInterpolation(context);
		} else if (s[0] == "<") {
			if (/[a-zA-Z]/.test(s[1])) {
				node = parseElement(context, ancesters);
			}
		}

		if (!node) {
			node = parseText(context);
		}

		nodes.push(node);
	}
	return nodes;
}

function isEnd(context: ParseContext, ancesters: ElementNode[]) {
	let s = context.source;
	if (s.startsWith("</")) {
		//? 为什么要遍历而不是直接检查栈顶？ 因为标签不一定是完全匹配的，因此要使用更健壮的检查策略，否则在出问题时会陷入死循环
		for (const el of ancesters) {
			if (s.slice(2, 2 + el.tag.length) === el.tag) return true;
		}
	}
	return !s;
}

//@fn parseText
function parseText(context: ParseContext): TextNode {
	let s = context.source;

	let endIndex = s.length;
	let endTokens = ["{{", "<"];
	let index = -1;
	for (const token of endTokens) {
		index = s.indexOf(token);
		if (index !== -1) {
			endIndex = Math.min(endIndex, index);
		}
	}

	const content = parseTextData(context, endIndex);

	return {
		type: NodeTypes.TEXT,
		content,
	};
}

//@fn parseTextData
function parseTextData(context: ParseContext, length: number) {
	const content = context.source.slice(0, length);
	advanceBy(context, length);
	return content;
}

//@fn parseElement
function parseElement(
	context: ParseContext,
	ancesters: ElementNode[]
): ElementNode {
	//? 思路，先提取出tag内容，然后进行 括号匹配？找到相应的闭合标签
	const element = parseTag(context, TagType.START);

	ancesters.push(element);
	element.children = parseChildren(context, ancesters);
	ancesters.pop();

	if (context.source.slice(2, 2 + element.tag.length) === element.tag) {
		parseTag(context, TagType.END);
	} else {
		throw new Error(element.tag + " lack end tag");
	}

	return element;
}

//@fn parseTag
function parseTag(context: ParseContext, type: TagType): ElementNode {
	const match = /^<\/?([a-z]+)/i.exec(context.source);
	let tag: string;
	if (match) {
		tag = match[1];
		advanceBy(context, match[0].length + 1);
	}

	if (type === TagType.END) return;

	return {
		type: NodeTypes.ELEMENT,
		tag,
		children: [],
	};
}

//@fn parseInterpolation
function parseInterpolation(context: ParseContext): InterPolationNode {
	//todo 解析插值
	//? {{exp}} -> exp
	const openDelimeter = "{{";
	const closeDelimeter = "}}";

	const closeIdx = context.source.indexOf(
		closeDelimeter,
		openDelimeter.length
	);

	advanceBy(context, openDelimeter.length); //舍弃前两位的 {{

	const contentLength = closeIdx - openDelimeter.length;
	const rawContent = parseTextData(context, contentLength);
	const content = rawContent.trim();

	advanceBy(context, closeDelimeter.length);

	return {
		type: NodeTypes.INTERPOLATION,
		content: {
			type: NodeTypes.SIMPLE_EXPRESSION,
			content,
		},
	};
}

function advanceBy(context: ParseContext, start: number) {
	context.source = context.source.slice(start);
}

function createRoot(children: ASTNode[]): ASTRoot {
	return {
		type: NodeTypes.ROOT,
		codegenNode: null,
		helpers: [],
		children,
	};
}

function createParserContext(content: string): ParseContext {
	return {
		source: content,
	};
}
