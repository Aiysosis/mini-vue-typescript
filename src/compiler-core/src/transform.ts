import {
	isElementNode,
	isInterpolationNode,
	isRootNode,
	isTextNode,
} from "./ast";
import { ASTNode, ASTRoot, ElementNode } from "./ast";
import {
	CREATE_ELEMENT_BLOCK,
	helperMapName,
	OPEN_BLOCK,
	TO_DISPLAY_STRING,
} from "./runtimeHelpers";

export type TransformPlugin = (node: ASTNode) => void;

export type TransformOptions = {
	nodeTransforms?: TransformPlugin[];
};

export type TransformContext = {
	root: ASTRoot;
	nodeTransforms: TransformPlugin[];
	helpers: Map<symbol, boolean>;
	helper: (name: symbol[]) => void;
};

export function transform(root: ASTRoot, options?: TransformOptions) {
	const context = createTransformContext(root, options);
	//*DFS
	traverseNode(root, context);

	createRootCodegen(root);

	root.helpers = [...context.helpers.keys()];
}

function createRootCodegen(root: ASTRoot) {
	root.codegenNode = root.children[0];
}

function createTransformContext(
	root: ASTRoot,
	options: TransformOptions
): TransformContext {
	const context: TransformContext = {
		root,
		nodeTransforms: options?.nodeTransforms || [],
		helpers: new Map(),
		helper(names) {
			for (const name of names) {
				context.helpers.set(name, true);
			}
		},
	};
	return context;
}

function traverseNode(node: ASTRoot | ElementNode, context: TransformContext) {
	const transforms = context.nodeTransforms;
	for (const child of node.children) {
		if (isRootNode(child)) {
			traverseNode(child, context);
		} else if (isElementNode(child)) {
			traverseNode(child, context);
			context.helper([OPEN_BLOCK, CREATE_ELEMENT_BLOCK]);
		} else if (isTextNode(child)) {
		} else if (isInterpolationNode(child)) {
			context.helper([TO_DISPLAY_STRING]);
		}

		for (const fn of transforms) {
			fn(child);
		}
	}
}
