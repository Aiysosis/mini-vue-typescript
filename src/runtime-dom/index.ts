import { Component } from "@/runtime-core/component";
import {
	createRenderer,
	Invoker,
	RendererNode,
	RendererOptions,
} from "@/runtime-core/renderer";

type DomElement = Element & {
	_vei?: Record<string, Invoker | undefined>;
};
type DomNode = Node;

const shouldSetAsProps = (el: DomElement, key: string, value: any) => {
	//特殊处理
	if (key === "form" && el.tagName === "input") return false;
	//兜底
	return key in el;
};
const domInterfaceImplement: RendererOptions<DomNode, DomElement> = {
	patchProp(el, key, prevValue, nextValue) {
		if (!nextValue) {
			el.removeAttribute(key);
		} else if (prevValue !== nextValue) {
			//* 分情况讨论
			const type = typeof el[key];

			//* 事件处理
			//? 约定以 on开头的属性都视为事件（onclick,onMouseover...）
			//? vue 中的 @click 之类的都是 onclick 的语法糖
			if (/^on/.test(key)) {
				//? 如果直接添加事件，删除事件会很麻烦，可以用一个变量把处理函数记录下来
				const invokers = el._vei || (el._vei = {}); //* vue event invoker
				const eventName = key.slice(2).toLowerCase();
				//? invoker 自身是一个函数，但是它的value属性才是真正的事件处理函数，自己相当于proxy
				let invoker = invokers[key];
				if (nextValue) {
					//* nextValue: Function | Function[]
					if (!invoker) {
						//* 缓存
						//invoker 是带有属性value的函数，它的value是缓存的事件处理函数
						invoker = ((e: any) => {
							//* 依次调用处理函数
							if (Array.isArray(invoker.value)) {
								for (const fn of invoker.value) {
									fn(e);
								}
							} else {
								invoker.value(e);
							}
						}) as Invoker;

						invoker.value = nextValue;
						//保存
						invokers[key] = invoker;
						el._vei = invokers;
						el.addEventListener(eventName, invoker);
					} else {
						//* 有 invoker 直接覆盖原有的
						invoker.value = nextValue;
					}
				} else if (invoker) {
					//* 如果没有处理函数，且invoker存在，那么要清空
					el.removeEventListener(eventName, invoker);
					//bug 不一定正确
					invokers[key] = undefined;
				}
			}
			//? class 使用el.className 兼容 + 提速
			else if (key === "class") {
				el.className = nextValue || "";
			} else if (shouldSetAsProps(el, key, nextValue)) {
				//* 要设置的属性是 DOM Properties
				//? 只有属性名的情况 例如 buttn 按钮的 disable
				if (type === "boolean" && nextValue === "") {
					el[key] = true;
				} else {
					//* 一般情形
					el[key] = nextValue;
				}
			} else {
				//* 设置的属性没有对应的 DOM Properties, 使用 setAttribute
				el.setAttribute(key, nextValue);
			}
		}
	},
	insert(el, parent, anchor?) {
		parent.insertBefore(el, anchor || null);
	},
	remove(el) {
		const parent = el.parentNode;
		if (parent) {
			parent.removeChild(el);
		}
	},
	createElement(type) {
		return document.createElement(type);
	},
	createText(text) {
		return document.createTextNode(text);
	},
	setText(node, text) {},
	setElementText(node, text) {
		node.textContent = text;
	},
};

export const renderer = createRenderer(domInterfaceImplement);

export function createApp(app: Component) {
	return renderer.createApp(app);
}

export * from "../runtime-core/index";
