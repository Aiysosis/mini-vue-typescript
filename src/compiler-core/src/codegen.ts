import { isTextNode } from "./ast";
import { ASTNode, ASTRoot } from "./parse";

export type CodegenContext = {
	code: string;
	push: (source: string) => void;
};

export function codegen(ast: ASTRoot) {
	const context = createCodegenContext();

	const { push } = context;

	const VueBinging = "vue";
	const helpers = ["toDisplayString"];
	const aliasHelpers = (s: string) => `${s} as _${s}`;
	push(`import { ${helpers.map(aliasHelpers)} } from "${VueBinging}"`);
	push("\n");

	push("export ");
	let functionName = "render";
	let args = ["_ctx", " _cache", "$props", "$setup", "$data", "$options"];
	const signature = args.join(", ");
	push(`function ${functionName}(${signature}){\n`);
	push("\treturn ");
	push("\n");
	genNode(ast.codegenNode, context);
	push("}");

	return {
		code: context.code,
	};
}
function genNode(node: ASTNode, context: CodegenContext) {
	const { push } = context;
	if (isTextNode(node)) {
		push(`"${node.content}"`);
	}
}
function createCodegenContext(): CodegenContext {
	const context = {
		code: "",
		push(source: string) {
			context.code += source;
		},
	};
	return context;
}
