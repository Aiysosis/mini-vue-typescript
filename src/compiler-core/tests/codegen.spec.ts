import { codegen } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";

describe("codegen", () => {
	// see https://template-explorer.vuejs.org
	test("string", () => {
		const ast = baseParse("hi");

		transform(ast);

		console.log("ast: ", ast);

		const { code } = codegen(ast);

		expect(code).toMatchSnapshot();
	});

	test.only("interpolation", () => {
		/**
         * *{{ message }}
         * import { toDisplayString as _toDisplayString } from "vue"

            export function render(_ctx, _cache, $props, $setup, $data, $options) {
            return _toDisplayString(_ctx.message)
            }
         */
		const ast = baseParse("{{ message }}");

		transform(ast);

		console.log("ast: ", ast);

		const { code } = codegen(ast);

		expect(code).toMatchSnapshot();
	});
});
