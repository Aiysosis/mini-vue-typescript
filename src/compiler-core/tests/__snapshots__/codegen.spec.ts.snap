// Vitest Snapshot v1

exports[`codegen > comprehension 1`] = `
"import { toDisplayString as _toDisplayString,openBlock as _openBlock,createElementBlock as _createElementBlock } from \\"vue\\"
export function render(_ctx,  _cache, $props, $setup, $data, $options){
	return (_openBlock, _createElementBlock(\\"div\\"), null, \\"hi,\\" + _toDisplayString(message))
}"
`;

exports[`codegen > element 1`] = `
"import { openBlock as _openBlock,createElementBlock as _createElementBlock } from \\"vue\\"
export function render(_ctx,  _cache, $props, $setup, $data, $options){
	return (_openBlock, _createElementBlock(\\"div\\"), null, )
}"
`;

exports[`codegen > interpolation 1`] = `
"import { toDisplayString as _toDisplayString } from \\"vue\\"
export function render(_ctx,  _cache, $props, $setup, $data, $options){
	return _toDisplayString(_ctx.message)
}"
`;

exports[`codegen > string 1`] = `
"
export function render(_ctx,  _cache, $props, $setup, $data, $options){
	return \\"hi\\"
}"
`;
